File: benchmarks/xlang/src/kmeans.py
import time
n = 200000; k = 10; iters = 15
xs = []; ys = []
seed = 42
for _ in range(n):
    seed = seed * 48271 % 2147483647
    xs.append(10.0 * seed / 2147483647.0)
    seed = seed * 48271 % 2147483647
    ys.append(10.0 * seed / 2147483647.0)
t0 = time.perf_counter()
stride = n // k
cx = [xs[c * stride] for c in range(k)]
cy = [ys[c * stride] for c in range(k)]
assign = [0] * n
for _ in range(iters):
    for i in range(n):
        best = 0; bd = 1000000.0
        xi = xs[i]; yi = ys[i]
        for c in range(k):
            dx = xi - cx[c]; dy = yi - cy[c]
            d = dx * dx + dy * dy
            if d < bd:
                bd = d; best = c
        assign[i] = best
    sx = [0.0] * k; sy = [0.0] * k; ct = [0] * k
    for i in range(n):
        c = assign[i]
        sx[c] += xs[i]; sy[c] += ys[i]; ct[c] += 1
    for c in range(k):
        if ct[c] > 0:
            cx[c] = sx[c] / ct[c]; cy[c] = sy[c] / ct[c]
sizes = [0] * k
for i in range(n):
    sizes[assign[i]] += 1
print(f"CHECK {max(sizes)}")
print(f"MS {int((time.perf_counter() - t0) * 1000)}")

File: benchmarks/xlang/src/kmeans.rs
fn main() {
    let n = 200000usize;
    let k = 10usize;
    let iters = 15;
    let mut xs = vec![0.0f64; n];
    let mut ys = vec![0.0f64; n];
    let mut seed = 42i64;
    for i in 0..n {
        seed = seed * 48271 % 2147483647;
        xs[i] = 10.0 * seed as f64 / 2147483647.0;
        seed = seed * 48271 % 2147483647;
        ys[i] = 10.0 * seed as f64 / 2147483647.0;
    }
    let t0 = std::time::Instant::now();
    let stride = n / k;
    let mut cx: Vec<f64> = (0..k).map(|c| xs[c * stride]).collect();
    let mut cy: Vec<f64> = (0..k).map(|c| ys[c * stride]).collect();
    let mut assign = vec![0usize; n];
    for _ in 0..iters {
        for i in 0..n {
            let mut best = 0;
            let mut bd = 1000000.0;
            for c in 0..k {
                let dx = xs[i] - cx[c];
                let dy = ys[i] - cy[c];
                let d = dx * dx + dy * dy;
                if d < bd { bd = d; best = c; }
            }
            assign[i] = best;
        }
        let mut sx = vec![0.0; k];
        let mut sy = vec![0.0; k];
        let mut ct = vec![0i64; k];
        for i in 0..n {
            let c = assign[i];
            sx[c] += xs[i]; sy[c] += ys[i]; ct[c] += 1;
        }
        for c in 0..k {
            if ct[c] > 0 { cx[c] = sx[c] / ct[c] as f64; cy[c] = sy[c] / ct[c] as f64; }
        }
    }
    let mut sizes = vec![0i64; k];
    for i in 0..n { sizes[assign[i]] += 1; }
    println!("CHECK {}", sizes.iter().max().unwrap());
    println!("MS {}", t0.elapsed().as_millis());
}
